sonic_platform_base/fan_base.py

```diff
@@ -66,6 +66,52 @@ def get_speed_tolerance(self):
         """
         raise NotImplementedError
 
+    def is_under_speed(self):
+        """
+        Calculates if the fan speed is under the tolerated low speed threshold
+
+        Default calculation requires get_speed_tolerance to be implemented, and checks
+        if the current fan speed (expressed as a percentage) is lower than <get_speed_tolerance>
+        percent below the target fan speed (expressed as a percentage)
+
+        Returns:
+            A boolean, True if fan speed is under the low threshold, False if not
+        """
+        speed = self.get_speed()
+        target_speed = self.get_target_speed()
+        tolerance = self.get_speed_tolerance()
+
+        for param, value in (('speed', speed), ('target speed', target_speed), ('speed tolerance', tolerance)):
+            if not isinstance(value, int):
+                raise TypeError(f'Fan {param} is not an integer value: {param}={value}')
+            if value < 0 or value > 100:
+                raise ValueError(f'Fan {param} is not a valid percentage value: {param}={value}')
+
+        return speed * 100 < target_speed * (100 - tolerance)
+
+    def is_over_speed(self):
+        """
+        Calculates if the fan speed is over the tolerated high speed threshold
+
+        Default calculation requires get_speed_tolerance to be implemented, and checks
+        if the current fan speed (expressed as a percentage) is higher than <get_speed_tolerance>
+        percent above the target fan speed (expressed as a percentage)
+
+        Returns:
+            A boolean, True if fan speed is over the high threshold, False if not
+        """
+        speed = self.get_speed()
```

Comment:
@spilkey-cisco This is an abstract class. Where is the HLD changes?

Reply:
My understanding is that https://github.com/sonic-net/sonic-platform-common/blob/master/sonic_fan/fan_base.py is the (obsolete, from what I can tell) fan abstract base class, and the file in this changeset follows the design from https://github.com/sonic-net/SONiC/blob/master/doc/platform_api/new_platform_api.md indicating ABCs are no longer used, 'No longer abstract base classes: All abstract methods simply raise "NotImplementedError"'. The new methods are not abstract since they have a default implementation, but can be overridden with vendor specific implementations if desired. If an HLD is needed, please advise on where it should go.